keps/sig-multicluster/2149-clusterid/README.md

```diff
@@ -0,0 +1,756 @@
+
+While a member of a ClusterSet, a cluster will also have an additional `clusterset.k8s.io ClusterClaim` which describes its current membership. This claim must be present exactly as long as the cluster's membership in a ClusterSet lasts, and removed when the cluster is no longer a member.
+
+More detail and examples of the uniqueness, lifespan, immutability, and content requirements for both the `id.k8s.io ClusterClaim` and `clusterset.k8s.io ClusterClaim` are described further below. The goal of these requirements are to provide to the MCS API a cluster ID of viable usefulness to address known user stories without being too restrictive or prescriptive.
+
+### User Stories
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### ClusterSet membership
+
+I have some set of clusters working together and need a way to uniquely identify them within the system that I use to track membership.
+
+#### Multi-Cluster Services
+
+I have a headless multi-cluster service deployed across clusters in my ClusterSet with similarly named pods in each cluster. I need a way to disambiguate each backend pod via DNS.
+
+```
+<<[UNRESOLVED]>>
+Examples of DNS using cluster ID a la `<hostname>.<clusterID>.<svc>.<ns>.svc.clusterset.local`
+<<[/UNRESOLVED]>>
+```
+
+#### Diagnostics
+
+Clusters within my ClusterSet send logs/metrics to a common monitoring solution and I need to be able to identify the cluster from which a given set of events originated.
+
+
+#### Joining or moving between ClusterSets
+
+I want the ability to add a previously-isolated cluster to a ClusterSet, or to move a cluster from one ClusterSet to another and be aware of this change.
+
+### `ClusterClaim` CRD
+```
+<<[UNRESOLVED]>>
+The actual name of the CRD is not finalized and is provisionally titled `ClusterClaim` for the remainder of this document.
+<<[/UNRESOLVED]>>
+```
+
+The `ClusterClaim` resource provides a way to store identification related, cluster scoped information for multi-cluster tools while creating flexibility for implementations. A cluster may have multiple `ClusterClaim`s, each holding a different identification related value. Each claim contains the following information:
+
+*   **Name** - a well known or custom name to identify the claim.
+*   **Value** - a claim-dependent string.
+
```

Comment:
I think it's worth noting here, up-front that the schema for ClusterClaim is intentionally loose because it is a multi-purpose tool, but certain well-known claims will add additional schema constraints, which may or may not be validated synchronously, depending on implementation.

Reply:
Also, we probably want some upper-bound on content size, e.g. 128KB

Reply:
Added both points (including stealing some of your prose) in b2aac28